### PR TITLE
Feature/compatibility hpos

### LINKED ISSUE #226: Incomatibility with High-Performance Order Storage (HPOS)
**Describe the bug**

The WooCommerce Amazon Pay plugin partially supports HPOS. We were able to test all the functionalities of the plugin except for integration with Subscriptions and Pre-Orders. Only found incombatibility are the Manual Actions in an Order's Edit screen (screenshot attached)

**To Reproduce**
1. [Enable](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#how-to-enable-hpos) HPOS
2. Complete an order using Amazon Pay and a simple product
3. Go to the completed order's edit screen
4. Manual actions added by the plugin aren't appearing

**Screenshots**

image (10)

**Expected behavior**
Manual Actions should be present and functional while using HPOS.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only the extension is active.
- [x] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by dpanta94**
### All Submissions:

* [x] Does your code follow the [Extendables](https://extendomattic.wordpress.com/standardizations/) standards?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changes proposed in this Pull Request:
Making Amazon Pay plugin compatible with HPOS

Made the plugin compatible using this [guide](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#declaring-extension-incompatibility)

Skipped:
- includes/compats/class-wc-amazon-payments-advanced-multi-currency-abstract.php L215-216 There is a comment present that mentions that sometimes get_post_meta could be needed. I didnt find any such case, but left it there anyway.

Closes #226 .

### How to test the changes in this Pull Request:

1. [Enable](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#how-to-enable-hpos) HPOS in your installation.
2. Plugins functionality should remain the same regardless of HPOS

### Other information:

* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changelog entry

> ADD - Compatibility with HPOS

### CODE REVIEW

File: includes/admin/legacy/class-wc-amazon-payments-advanced-order-admin-legacy.php [hunk 1]
```diff
@@ -42,8 +42,9 @@ public function do_order_action( $order, $id, $action, $version ) {
 				$this->clear_stored_states( $order_id );
 				break;
 			case 'authorize':
-				delete_post_meta( $order_id, 'amazon_authorization_id' );
-				delete_post_meta( $order_id, 'amazon_capture_id' );
+				$order->delete_meta_data( 'amazon_authorization_id' );
```
Comment (kalessil) [inline includes/admin/legacy/class-wc-amazon-payments-advanced-order-admin-legacy.php:45]: Will this be compatible with older supported WooCommerce versions?

File: includes/class-wc-gateway-amazon-payments-advanced-privacy.php [hunk 2]
```diff
@@ -292,14 +291,16 @@ protected function maybe_handle_order( $order ) {
...
+			$meta_value = $order->get_meta( $key, true, 'edit' );
 			if ( empty( $meta_value ) ) {
 				continue;
 			}
-			delete_post_meta( $order_id, $key );
+			$order->delete_meta_data( $key );
 			$deleted = true;
 		}
 
+		$order->save();
```
Comment (kalessil) [inline includes/class-wc-gateway-amazon-payments-advanced-privacy.php:302]: Why?

File: includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php [hunk 3]
```diff
@@ -1075,6 +1104,10 @@ public static function handle_payment_authorization_response( $response, $order,
 	 * @return bool Returns true if succeed.
 	 */
 	public static function update_order_from_authorize_response( $order, $response, $capture_now = false ) {
+		if ( ! ( $order instanceof \WC_Order ) ) {
```
Comment (kalessil) [inline includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php:1107]: This is a bot controversial towards what's stated in doc-block.

File: includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php [hunk 4]
```diff
@@ -574,23 +593,29 @@ public static function get_capture_state( $order_id, $id ) {
 	public static function get_order_ref_state( $order_id, $state = 'amazon_reference_state' ) {
 		$ret_state = '';
 
+		$or = wc_get_order( $order_id );
```
Comment (kalessil) [inline includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php:596]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 5]
```diff
@@ -631,14 +625,14 @@ protected function get_missing_capture_id_request_error( $order_post ) {
 	 * @return WP_Post|WP_Error Post object if it's valid, WP_Error if it's invalid.
 	 */
 	protected function is_valid_order( $order_id ) {
-		$order_post = get_post( (int) $order_id );
+		$or = wc_get_order( (int) $order_id );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:628]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 6]
```diff
@@ -615,7 +609,7 @@ protected function get_missing_authorization_id_request_error( $order_post ) {
 	 * @return null|WP_Error Null if there's no error in the request.
 	 */
 	protected function get_missing_capture_id_request_error( $order_post ) {
-		$ref_id = get_post_meta( $order_post->ID, 'amazon_capture_id', true );
+		$ref_id = $order_post->get_meta( 'amazon_capture_id', true, 'edit' );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:612]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 7]
```diff
@@ -599,7 +593,7 @@ protected function get_missing_reference_id_request_error( $order_post ) {
 	 * @return null|WP_Error Null if there's no error in the request.
 	 */
 	protected function get_missing_authorization_id_request_error( $order_post ) {
-		$ref_id = get_post_meta( $order_post->ID, 'amazon_authorization_id', true );
+		$ref_id = $order_post->get_meta( 'amazon_authorization_id', true, 'edit' );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:596]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 8]
```diff
@@ -583,7 +577,7 @@ public function refund( $request ) {
 	 * @return null|WP_Error Null if there's no error in the request.
 	 */
 	protected function get_missing_reference_id_request_error( $order_post ) {
-		$ref_id = get_post_meta( $order_post->ID, 'amazon_reference_id', true );
+		$ref_id = $order_post->get_meta( 'amazon_reference_id', true, 'edit' );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:580]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 9]
```diff
@@ -522,33 +518,31 @@ public function capture( $request ) {
 	 *                                   WP_REST_Response instance.
 	 */
 	public function refund( $request ) {
-		$order_post = $this->is_valid_order( $request['order_id'] );
-		if ( is_wp_error( $order_post ) ) {
-			return $order_post;
+		$or = $this->is_valid_order( $request['order_id'] );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:521]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 10]
```diff
@@ -463,49 +462,46 @@ public function close_authorization( $request ) {
 	 *                                   WP_REST_Response instance.
 	 */
 	public function capture( $request ) {
-		$order_post = $this->is_valid_order( $request['order_id'] );
-		if ( is_wp_error( $order_post ) ) {
-			return $order_post;
+		$or = $this->is_valid_order( $request['order_id'] );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:465]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 11]
```diff
@@ -415,22 +416,20 @@ protected function authorize_order_v1( $order_id, $authorize_args = array() ) {
 	 *                                   WP_REST_Response instance.
 	 */
 	public function close_authorization( $request ) {
-		$order_post = $this->is_valid_order( $request['order_id'] );
-		if ( is_wp_error( $order_post ) ) {
-			return $order_post;
+		$or = $this->is_valid_order( $request['order_id'] );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:419]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 12]
```diff
@@ -415,22 +416,20 @@ protected function authorize_order_v1( $order_id, $authorize_args = array() ) {
...
-			$error = $this->get_missing_authorization_id_request_error( $order_post );
+			$error = $this->get_missing_authorization_id_request_error( $or );
 			if ( is_wp_error( $error ) ) {
 				return $error;
 			}
 
-			$order_id = (int) $request['order_id'];
-			$auth_id  = get_post_meta( $order_id, 'amazon_authorization_id', true );
-			$resp     = WC_Amazon_Payments_Advanced_API_Legacy::close_authorization( $order_id, $auth_id );
+			$auth_id = $or->get_meta( 'amazon_authorization_id', true, true );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:431]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 13]
```diff
@@ -415,22 +416,20 @@ protected function authorize_order_v1( $order_id, $authorize_args = array() ) {
...
+			$resp    = WC_Amazon_Payments_Advanced_API_Legacy::close_authorization( $or->get_id(), $auth_id );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:432]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 14]
```diff
@@ -384,16 +383,18 @@ protected function authorize_order_v1( $order_id, $authorize_args = array() ) {
 			return $resp;
 		}
 
+		$or = wc_get_order( $order_id );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:386]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 15]
```diff
@@ -337,29 +336,29 @@ public function authorize( $request ) {
 	 *                                   WP_REST_Response instance.
 	 */
 	public function authorize_and_capture( $request ) {
-		$order_post = $this->is_valid_order( $request['order_id'] );
-		if ( is_wp_error( $order_post ) ) {
-			return $order_post;
+		$or = $this->is_valid_order( $request['order_id'] );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:339]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 16]
```diff
@@ -299,28 +298,28 @@ public function get_reference_state( $request ) {
 	 *                                   WP_REST_Response instance.
 	 */
 	public function authorize( $request ) {
-		$order_post = $this->is_valid_order( $request['order_id'] );
-		if ( is_wp_error( $order_post ) ) {
-			return $order_post;
+		$or = $this->is_valid_order( $request['order_id'] );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:301]: Can we give this variable a better name, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 17]
```diff
@@ -246,39 +246,38 @@ protected function get_permissions_check_error_message( $action ) {
 	 * @return array|WP_Error|WP_REST_Response
 	 */
 	public function get_reference_state( $request ) {
-		$order_post = $this->is_valid_order( $request['order_id'] );
+		$or = $this->is_valid_order( $request['order_id'] );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:249]: Can we give this variable a better name, please?

File: includes/class-wc-gateway-amazon-payments-advanced-privacy.php (hunk 2 above)
Comment (dpanta94) [reply includes/class-wc-gateway-amazon-payments-advanced-privacy.php:302]: there is a foreach a couple of lines above which uses WC_Order::delete_meta_data. we need to commit the transaction with save then.

File: includes/admin/legacy/class-wc-amazon-payments-advanced-order-admin-legacy.php (hunk 1 above)
Comment (dpanta94) [reply includes/admin/legacy/class-wc-amazon-payments-advanced-order-admin-legacy.php:45]: Yea we require at least WC 4.0 right now and that method was introduced in WC 2.6.0 as stated in woocommerce/includes/abstracts/abstract-wc-data.php WC_Data::delete_meta_data doc comment

File: woocommerce-gateway-amazon-payments-advanced.php [hunk 18]
```diff
@@ -504,19 +514,26 @@ public function rest_api_register_routes() {
...
-				'amazon_authorization_id'    => get_post_meta( $post->ID, 'amazon_authorization_id', true ),
-				'amazon_capture_state'       => WC_Amazon_Payments_Advanced_API_Legacy::get_order_ref_state( $post->ID, 'amazon_capture_state' ),
-				'amazon_capture_id'          => get_post_meta( $post->ID, 'amazon_capture_id', true ),
-				'amazon_refund_ids'          => get_post_meta( $post->ID, 'amazon_refund_id', false ),
-			);
+		if ( 'amazon_payments_advanced' !== $response->data['payment_method'] ) {
+			return $response;
 		}
 
+		$or = wc_get_order( $post->ID );
```
[the remark this answers is not in the document]
Comment (dpanta94) [reply woocommerce-gateway-amazon-payments-advanced.php]: done!

File: includes/admin/class-wc-amazon-payments-advanced-order-admin.php [hunk 19]
```diff
@@ -42,7 +42,15 @@ public function order_actions() {
 	 * Non AJAX handler that performs order actions.
 	 */
 	public function order_actions_non_ajax() {
-		if ( 'shop_order' !== get_current_screen()->id ) {
+		if ( function_exists( 'wc_get_container' ) && class_exists( 'Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController' ) ) {
```
[the remark this answers is not in the document]
Comment (dpanta94) [reply includes/admin/class-wc-amazon-payments-advanced-order-admin.php]: done!

File: includes/admin/class-wc-amazon-payments-advanced-order-admin.php [hunk 20]
```diff
@@ -52,8 +60,20 @@ public function order_actions_non_ajax() {
 
 		check_admin_referer( 'amazon_order_action', 'security' );
 
-		$order_id = absint( $_GET['post'] ); // TODO: This may break when custom data stores are implemented in the future.
-		$order    = wc_get_order( $order_id );
+		$order_id = ! empty( $_GET['post'] ) ? $_GET['post'] : false;
```
[the remark this answers is not in the document]
Comment (dpanta94) [reply includes/admin/class-wc-amazon-payments-advanced-order-admin.php:58]: done!

File: includes/admin/class-wc-amazon-payments-advanced-order-admin.php [hunk 21]
```diff
@@ -104,37 +124,53 @@ public function do_order_action( $order, $id, $action, $version ) {
...
+		if ( ! ( $order instanceof \WC_Order ) ) {
 			return;
 		}
 
 		if ( 'amazon_payments_advanced' !== wc_apa_get_order_prop( $order, 'payment_method' ) ) {
 			return;
 		}
 
-		$post_types = apply_filters( 'woocommerce_amazon_pa_admin_meta_box_post_types', array( 'shop_order' ) );
+		if ( function_exists( 'wc_get_container' ) && class_exists( 'Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController' ) ) {
```
[the remark this answers is not in the document]
Comment (dpanta94) [reply includes/admin/class-wc-amazon-payments-advanced-order-admin.php]: done!

File: includes/admin/class-wc-amazon-payments-advanced-order-admin.php [hunk 22]
```diff
@@ -104,37 +124,53 @@ public function do_order_action( $order, $id, $action, $version ) {
...
+				$post_type,
+				'side'
+			);
 		}
 	}
 
 	/**
 	 * Authorization metabox content.
+	 *
+	 * @param Object $object The current post/order object.
```
[the remark this answers is not in the document]
Comment (dpanta94) [reply includes/admin/class-wc-amazon-payments-advanced-order-admin.php]: done!

File: includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php (hunk 4 above)
Comment (dpanta94) [reply includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 5 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 9 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 10 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 11 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 14 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 15 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 7 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 12 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 13 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 6 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

**Comment by dpanta94** [comment]
Hey @kalessil ! Amended/answered to all your comments

File: includes/class-wc-gateway-amazon-payments-advanced-privacy.php (hunk 2 above)
Comment (kalessil) [reply includes/class-wc-gateway-amazon-payments-advanced-privacy.php:302]: I see, thanks for explaining!

File: includes/admin/legacy/class-wc-amazon-payments-advanced-order-admin-legacy.php (hunk 1 above)
Comment (kalessil) [reply includes/admin/legacy/class-wc-amazon-payments-advanced-order-admin-legacy.php:45]: I see, thanks for explaining!

File: includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php (hunk 3 above)
Comment (kalessil) [reply includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php]: Thanks!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 23]
```diff
@@ -188,11 +188,11 @@ protected function get_permissions_check( $request, $action ) {
 				break;
 			case 'delete':
 			case 'edit':
-				$post           = get_post( (int) $request['order_id'] );
+				$post           = wc_get_order( (int) $request['order_id'] );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:191]: Here we should probably rename the variable. WDYT?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 24]
```diff
@@ -583,8 +577,8 @@ public function refund( $request ) {
 	 * @return null|WP_Error Null if there's no error in the request.
 	 */
 	protected function get_missing_reference_id_request_error( $order_post ) {
-		$ref_id = get_post_meta( $order_post->ID, 'amazon_reference_id', true );
-		if ( ! $ref_id ) {
+		$reference_id = $order_post->get_meta( 'amazon_reference_id', true, 'edit' );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:580]: It looks like the doc-block for `$order_post` is outdated, + we probably could give it a better name. WDYT?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 25]
```diff
@@ -599,8 +593,8 @@ protected function get_missing_reference_id_request_error( $order_post ) {
 	 * @return null|WP_Error Null if there's no error in the request.
 	 */
 	protected function get_missing_authorization_id_request_error( $order_post ) {
-		$ref_id = get_post_meta( $order_post->ID, 'amazon_authorization_id', true );
-		if ( ! $ref_id ) {
+		$reference_id = $order_post->get_meta( 'amazon_authorization_id', true, 'edit' );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:596]: Similar here: doc-block + naming of `$order_post`.

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 26]
```diff
@@ -615,8 +609,8 @@ protected function get_missing_authorization_id_request_error( $order_post ) {
 	 * @return null|WP_Error Null if there's no error in the request.
 	 */
 	protected function get_missing_capture_id_request_error( $order_post ) {
-		$ref_id = get_post_meta( $order_post->ID, 'amazon_capture_id', true );
-		if ( ! $ref_id ) {
+		$reference_id = $order_post->get_meta( 'amazon_capture_id', true, 'edit' );
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:612]: Similar here: doc-block + naming of `$order_post`.

File: includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php [hunk 27]
```diff
@@ -574,23 +593,29 @@ public static function get_capture_state( $order_id, $id ) {
 	public static function get_order_ref_state( $order_id, $state = 'amazon_reference_state' ) {
 		$ret_state = '';
```
Comment (kalessil) [inline includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php:594]: Can we improve the name of this variable, please?

File: includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php [hunk 28]
```diff
@@ -574,23 +593,29 @@ public static function get_capture_state( $order_id, $id ) {
...
+		$order = wc_get_order( $order_id );
+
+		if ( ! ( $order instanceof \WC_order ) ) {
+			return $ret_state;
+		}
+
 		switch ( $state ) {
 			case 'amazon_reference_state':
-				$ref_id = get_post_meta( $order_id, 'amazon_reference_id', true );
+				$ref_id = $order->get_meta( 'amazon_reference_id', true, 'edit' );
```
Comment (kalessil) [inline includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php:604]: Also variable name.

**Review by kalessil — CHANGES_REQUESTED** [review]
Thanks for pushing new changes @dpanta94 - it's easier to review the changes now.
Looks good overall, I leaving new comments and will make a final review round when you are ready.

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 23 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 24 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 25 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 26 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php]: done!

File: includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php (hunk 27 above)
Comment (dpanta94) [reply includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php]: done!

File: includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php (hunk 28 above)
Comment (dpanta94) [reply includes/legacy/class-wc-amazon-payments-advanced-api-legacy.php]: done!

**Comment by dpanta94** [comment]
Hey @kalessil ! amended your latest comments ! :D

File: includes/admin/legacy/class-wc-amazon-payments-advanced-order-admin-legacy.php [hunk 29]
```diff
@@ -108,9 +110,15 @@ public function do_order_action( $order, $id, $action, $version ) {
...
-		delete_post_meta( $order_id, 'amazon_authorization_state' );
+		$order = wc_get_order( $order_id );
+		if ( ! ( $order instanceof WC_Order ) ) {
+			return;
+		}
+
+		$order->delete_meta_data( 'amazon_reference_state' );
+		$order->delete_meta_data( 'amazon_capture_state' );
+		$order->delete_meta_data( 'amazon_authorization_state' );
+		$order->save();
```
Comment (kalessil) [inline includes/admin/legacy/class-wc-amazon-payments-advanced-order-admin-legacy.php:121]: Nitpick: can we add a new line after save, please?

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php [hunk 30]
```diff
@@ -246,20 +246,19 @@ protected function get_permissions_check_error_message( $action ) {
...
 		// If `refresh=1` is passed, cache ref states will be cleared so it
 		// makes a call to Amazon to get the details.
 		if ( ! empty( $request['refresh'] ) ) {
-			delete_post_meta( $order_post->ID, 'amazon_reference_state' );
-			delete_post_meta( $order_post->ID, 'amazon_capture_state' );
-			delete_post_meta( $order_post->ID, 'amazon_authorization_state' );
+			$order->delete_meta_data( 'amazon_reference_state' );
+			$order->delete_meta_data( 'amazon_capture_state' );
+			$order->delete_meta_data( 'amazon_authorization_state' );
+			$order->save();
```
Comment (kalessil) [inline includes/class-wc-amazon-payments-advanced-rest-api-controller.php:261]: Also a new line, please.

File: includes/legacy/class-wc-gateway-amazon-payments-advanced-legacy.php [hunk 31]
```diff
@@ -441,7 +443,8 @@ protected function process_payment_capture( $order, $amazon_reference_id ) {
 	 */
 	public function process_async_auth( $order, $amazon_reference_id ) {
 
-		update_post_meta( $order->get_id(), 'amazon_timed_out_transaction', true );
+		$order->update_meta_data( 'amazon_timed_out_transaction', true );
+		$order->save();
```
Comment (kalessil) [inline includes/legacy/class-wc-gateway-amazon-payments-advanced-legacy.php:447]: Can we add a new line please?

**Review by kalessil — APPROVED** [review]
LGTM! I left minor comments, it'd be great to have them addressed.

File: includes/admin/legacy/class-wc-amazon-payments-advanced-order-admin-legacy.php (hunk 29 above)
Comment (dpanta94) [reply includes/admin/legacy/class-wc-amazon-payments-advanced-order-admin-legacy.php:121]: done!

File: includes/class-wc-amazon-payments-advanced-rest-api-controller.php (hunk 30 above)
Comment (dpanta94) [reply includes/class-wc-amazon-payments-advanced-rest-api-controller.php:261]: done!

File: includes/legacy/class-wc-gateway-amazon-payments-advanced-legacy.php (hunk 31 above)
Comment (dpanta94) [reply includes/legacy/class-wc-gateway-amazon-payments-advanced-legacy.php:447]: done!

**Pull request merged**